### DISCUSSION TITLE
NPM version lookups failing on private NPM registry due to auth

### DISCUSSION THREAD

**Discussion by rvillane** [renovatebot/renovate#22367] [Request Help] [answered]
### I'm using the Mend Renovate hosted app on github.com.

- [ ] I'm using the Mend Renovate hosted app.

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe your problem.

Hello, I'm transitioning from a self-hosted Nexus to Google Artifact Registry (GAR) as private NPM registry.  Nexus was fully open to read requests and anonymous requests were allowed, hence renovate had an easy job to run the NPM version lookups, however GAR requires access tokens.

After login to GAR, the ephemeral **_authToken** (lives 60 min) access token is appended to .npmrc file and I'm attempting to programmatically read the content of this .npmrc file to populate the `config.npmrc` field

**.npmrc:**
```
registry=https://us-npm.pkg.dev/my-project/my-private-registry-npm/
//us-npm.pkg.dev/my-project/my-private-registry-npm/:always-auth=true
//us-npm.pkg.dev/my-project/my-private-registry-npm/:_authToken="ya29.c.b0A..............."
```

**config.js:**
```
const fs = require ('fs');
const npmrc = fs.readFileSync('/tmp-auth/.npmrc','utf8');

module.exports = {
  endpoint: 'https://my.ghe.com/api/v3/',
  platform: 'github',
  onboarding: true,
  onboardingConfig: {
    extends: ['config:base'],
  },
  autodiscover: true,
  autodiscoverFilter: '************',
  requireConfig: "required",
  automerge: false,
  extends: ['group:allNonMajor'],
  enabledManagers: ['npm','dockerfile'],
  rangeStrategy: 'pin',
  recreateClosed: true,
  minimumReleaseAge: "7 days",
  prCreation:'not-pending',
  packageRules: [
    {
      matchDatasources: ['docker'],
      prCreation: 'immediate',
    },
    {
      "matchPackageNames": ["node"],
      "allowedVersions": "<19"
    },    
  ],
  lockFileMaintenance: { enabled: true },
  bumpVersion: 'patch',
  npmrc
};
```

But NPM version lookups are failing with HTTP 403 error code as follow:

```
DEBUG: GET https://us-npm.pkg.dev/my-project/my-private-registry-npm/coverify = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=403 retryCount=0, duration=197) (repository=MyOrg/myRepo)
DEBUG: Failed to look up npm package coverify (repository=MyOrg/myRepo, packageFile=package.json, dependency=coverify)
DEBUG: GET https://us-npm.pkg.dev/my-project/my-private-registry-npm/@google-cloud%2Fiam-credentials = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=403 retryCount=0, duration=223) (repository=MyOrg/myRepo)
DEBUG: Failed to look up npm package @google-cloud/iam-credentials (repository=MyOrg/myRepo, packageFile=package.json, dependency=@google-cloud/iam-credentials)
DEBUG: GET https://us-npm.pkg.dev/my-project/my-private-registry-npm/moment = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=403 retryCount=0, duration=228) (repository=MyOrg/myRepo)
DEBUG: Failed to look up npm package moment (repository=MyOrg/myRepo, packageFile=package.json, dependency=moment)
DEBUG: GET https://us-npm.pkg.dev/my-project/my-private-registry-npm/properties-reader = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=403 retryCount=0, duration=260) (repository=MyOrg/myRepo)
DEBUG: Failed to look up npm package properties-reader (repository=MyOrg/myRepo, packageFile=package.json, dependency=properties-reader)
DEBUG: GET https://us-npm.pkg.dev/my-project/my-private-registry-npm/eslint-plugin-import = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=403 retryCount=0, duration=152) (repository=MyOrg/myRepo)
DEBUG: Failed to look up npm package eslint-plugin-import (repository=MyOrg/myRepo, packageFile=package.json, dependency=eslint-plugin-import)
DEBUG: GET https://us-npm.pkg.dev/my-project/my-private-registry-npm/eslint-config-airbnb = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=403 retryCount=0, duration=165) (repository=MyOrg/myRepo)
```

I'm scratching my head why the NPM lookup's are failing due to auth issues if the access token is available and I'm wondering if the style to populate the `config.npmrc` field might be causing those problems, according to [1], new lines should be replaced with "\n" but I'm not clear how strict is renovate about newlines and the DEBUG logs are masking the config.npmrc content due to security reasons I suppose then is not totally clear what is the final content of the config.js file.

[1] https://docs.renovatebot.com/configuration-options/#npmrc

```
DEBUG: File config
       "config": {
         "endpoint": "https://my.ghe.com/api/v3/",
         "platform": "github",
         "onboarding": true,
         "onboardingConfig": {"extends": ["config:base"]},
         "autodiscover": true,
         "autodiscoverFilter": "*******)",
         "requireConfig": "required",
         "automerge": false,
         "extends": ["group:allNonMajor"],
         "enabledManagers": ["npm", "dockerfile"],
         "rangeStrategy": "pin",
         "recreateClosed": true,
         "minimumReleaseAge": "7 days",
         "prCreation": "not-pending",
         "packageRules": [
           {"matchDatasources": ["docker"], "prCreation": "immediate"},
           {"matchPackageNames": ["node"], "allowedVersions": "<19"}
         ],
         "lockFileMaintenance": {"enabled": true},
         "bumpVersion": "patch",
         "npmrc": "***********"
       }
DEBUG: CLI config
       "config": {}
DEBUG: Env config
       "config": {
         "hostRules": [],
         "token": "***********",
         "gitAuthor": "Renovate Bot <email@myserver.com>"
       }
DEBUG: Combined config
       "config": {
         "endpoint": "https://my.ghe.com/api/v3/",
         "platform": "github",
         "onboarding": true,
         "onboardingConfig": {"extends": ["config:base"]},
         "autodiscover": true,
         "autodiscoverFilter": "************",
         "requireConfig": "required",
         "automerge": false,
         "extends": ["group:allNonMajor"],
         "enabledManagers": ["npm", "dockerfile"],
         "rangeStrategy": "pin",
         "recreateClosed": true,
         "minimumReleaseAge": "7 days",
         "prCreation": "not-pending",
         "packageRules": [
           {"matchDatasources": ["docker"], "prCreation": "immediate"},
           {"matchPackageNames": ["node"], "allowedVersions": "<19"}
         ],
         "lockFileMaintenance": {"enabled": true},
         "bumpVersion": "patch",
         "npmrc": "***********",
         "hostRules": [],
         "token": "***********",
         "gitAuthor": "Renovate Bot <email@myserver.com>"
       }
DEBUG: Adding trailing slash to endpoint
DEBUG: Found valid git version: 2.40.1
DEBUG: Detected GitHub Enterprise Server, version: 3.6.12
DEBUG: Platform config
       "platformConfig": {
         "hostType": "github",
         "endpoint": "https://my.ghe.com/api/v3/",
         "isGHApp": false,
         "isGhe": true,
         "gheVersion": "3.6.12",
         "userDetails": {
           "username": "renovate-user",
           "name": "Service Renovate User"
         }
       },
       "renovateUsername": "renovate-user"
```

Any ideas or recommendations ?  Any obvious flaws that you can notice in the configuration described above ?

thanks in advance

### Relevant debug logs.

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by rarkins** [comment]
A couple of quick things:
- Can you try one run where you've hardcoded the `npmrc` value into into your `config.js`, to rule out whether the file is being read correctly in the first place?
- Another way to confirm that partially is to put an invalid URL into the registry= to make sure it's going to the wrong place
- Can you remove all auth and see if the error remains the same? Trying to confirm that a token is being sent at all
- Try doing a regex replace on the content after you read it, e.g. replace `\n` with `\\n`

**Reply by rvillane** [reply]
@rakins , thanks a lot for the suggestions, below is my initial round of responses:

1- I tried a hardcoded `config.npmrc` value with 2 values below, but I've got the same HTTP 403 error code back when running NPM lookups:
- npmrc: "registry=https://us-npm.pkg.dev/my-project/my-private-registry-npm/\\n//us-npm.pkg.dev/my-project/my-private-registry-npm/:always-auth=true\\n//us-npm.pkg.dev/my-project/my-private-registry-npm/:_authToken=ya29.c.b0Aaekm1Lib7........."
- npmrc: "registry=https://us-npm.pkg.dev/my-project/my-private-registry-npm/\n//us-npm.pkg.dev/my-project/my-private-registry-npm/:always-auth=true\n//us-npm.pkg.dev/my-project/my-private-registry-npm/:_authToken=ya29.c.b0Aaekm1Lib7........."

2- Tried the invalid registry URL, but it caused the GHE repo to not be scanned by the bot, at least I found no trace of scanned npm apps in logs.  As soon as the registry URL was fixed, the GHE repos with npm manager showed again in the bot debug logs.
3- removing `_authToken` field caused the same HTTP 403 error back when running NPM lookups.
4- Tried the `\n` replacement in 1) as part of the hardcoded npmrc value

**Reply by rarkins** [reply]
I have debugged using `config.js` with your example rule pasted in, with new lines converted to `\n`.

This looks correct:
<img width="573" alt="image" src="https://github.com/renovatebot/renovate/assets/6311784/d5f9f123-666a-4b34-b01e-8fd5f5d214ae">

The queried URL will be like `https://us-npm.pkg.dev/my-project/my-private-registry-npm/left-pad`

At the lowest level I see the `token` applied:

<img width="413" alt="image" src="https://github.com/renovatebot/renovate/assets/6311784/76edc123-d219-4427-947c-12351ca31270">

This seems to confirm that the config.npmrc has been read and used as intended.

So I think the next step is for you to confirm:
- Is the URL right?
- Does curling that URL with the same Authorization Bearer approach (but your real token) work?

**Reply by rvillane** [reply]
I tried curling several NPM public libraries using the suggested approach and it works fine:

`curl -H 'authorization: Bearer ya29.c.b0Aaekm1......'   https://us-npm.pkg.dev/my-project/my-private-registry-npm/left-pad`

then yes, the URL is correct (is probably obvious, but the one above is not the real registry URL).

However, if I use that exact same authToken and put it in the hardcoded `config.npmrc` field with `\n` separator, NPM lookups results in HTTP 403

`npmrc: "registry=https://us-npm.pkg.dev/my-project/my-private-registry-npm/\n//us-npm.pkg.dev/my-project/my-private-registry-npm/:always-auth=true\n//us-npm.pkg.dev/my-project/my-private-registry-npm/:_authToken=ya29.c.b0Aaekm1......"`

**Reply by rarkins** [reply]
Unfortunately I'm out of ideas on this - it would need you to run debug from source code in your environment

**Comment by rvillane** [comment]
Hi @rarkins, I was able to reproduce the issue on my local environment.  My findings below:

1. NPM datasource setNpmrc() function compares the `config.npmrc` field against the npmrc in the `npmrc`'s  repo renovate.json file, since npmrc field is missing in the repo renovate.json file is the same value and the rest of the function is not executed, I suspect this behavior is intended -> https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/npm/npmrc.ts#L127 
2. NPM datasource getDependency() function where the NPM lookup happens uses an HTTP object with the auth credentials set to `null`, `config.npmrc` settings were never used, running anonymous NPM lookups. The `options` object passed as param is empty. See the following links:

- https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/npm/index.ts#L30
- https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/npm/get.ts#L120

I'm not totally confident this is the actual issue as any other renovate users dealing with a private NPM registry enforcing authentication would be running into this same issue, but this is what I have found so far.

![renovate-debug](https://github.com/renovatebot/renovate/assets/11447528/e22c860e-870a-422a-a9a4-582ae7f40f62)

**Reply by rarkins** [reply]
Is there any difference if you set npmrcMerge=true?

**Reply by rvillane** [reply]
since I'm running this on my local environment, my $HOME/.npmrc has the legacy NPM registry I'm migrating away from, adding more noise to the mix.

**Comment by viceice** [comment]
authentication is handled a layer lower on http class, so it's expected to not have any auth inside datasource.

renovate should read the auth at npm manager extract phase and configure hostrule there.

I'm working with private npm package since years and don't have any issues when directly configure auth via host rule instead of npmrc.
so I only have the registry url inside the repo npmrc file and renovate add the auth token on demand from hostrule.

**Reply by rvillane** [reply]
Hi @viceice , thanks for your response, I dug deeper into the code and I see what do you mean, found the following references where the GOT library is used to make the actual HTTP calls and I can confirm the authorization header containing the Bearer token is passed:

- https://github.com/renovatebot/renovate/blob/main/lib/util/http/index.ts#L172
- https://github.com/renovatebot/renovate/blob/main/lib/util/http/index.ts#L91-L92

Also, based on what I see in the following screenshot, seems like the NPM lookups are actually succeeding, now I'm scratching my head why the NPM lookups are reported as failed with an HTTP 403 error

![got-successful-responses](https://github.com/renovatebot/renovate/assets/11447528/3b5e7f5b-7944-4ef9-802d-e8f0b655c736)

**Reply by rarkins** [reply]
Is it failing during release notes retrieval? You could set `fetchReleaseNotes=false` to confirm that

**Reply by rvillane** [reply]
Seems to be during the NPM lookup, the first requests are succeeding but then they all start to fail with 403 error

![npm-lookup-403](https://github.com/renovatebot/renovate/assets/11447528/f0a392c3-e1a7-4a4b-9def-bd79b7343fcf)

**Reply by rvillane** [reply]
@rarkins I noticed renovate uses a caching layer in the HTTP class:

https://github.com/renovatebot/renovate/blob/main/lib/util/http/index.ts#L137-L193

and would like to discard cached content as the root cause of the issue, is there a config setting to suppress caching ?

**Reply by rarkins** [reply]
Low level http caching can't be disabled, as far as I'm aware. But if the early request worked, you'd expect the cached result to be good?

**Reply by rvillane** [reply]
right, cached result should be good, I'm just getting out of ideas why it works initially and suddenly starts failing with 403 error

**Reply by rarkins** [reply]
So the package is in multiple package files in sane repo? Or fails for packages in only one folder too?

**Comment by rvillane** [comment]
With the additional debugging context, I ran the bot again and looking at the log file (enclosed), the first GHE repo works fine and NPM lookups succeed, and then the other 2 subsequent GHE repos fail the NPM lookups.

1) When the first GHE repo was scanned, the following log statements were printed, it seems to reset the hostRules and add host rules and authentication settings for the GHE server and Artifact Registry repository.  All this is expected and working fine

```
"Reinitializing hostRules for repo","time":"2023-05-29T16:21:54.479Z"
"Clearing hostRules","time":"2023-05-29T16:21:54.479Z"
"Adding token authentication for my-ghe-server.com to hostRules"
"repository":"MyGHEOrg/token-signer","renovateVersion":"35.98.0","msg":"Repository started"
"repository":"MyGHEOrg/token-signer","msg":"PackageFiles.clear() - Package files deleted"
"repository":"MyGHEOrg/token-signer","msg":"initRepo(\"MyGHEOrg/token-signer\")"
"repository":"MyGHEOrg/token-signer","msg":"MyGHEOrg/token-signer default branch = main"
"repository":"MyGHEOrg/token-signer","msg":"Using personal access token for git init"
"repository":"MyGHEOrg/token-signer","msg":"Setting npmrc"
"repository":"MyGHEOrg/token-signer","msg":"Adding token authentication for https://us-npm.pkg.dev/my-proj/my-private-npm/ to hostRules"
```

2) When the bot transitions from the first repo into the second repo, I found the following logs were printed.  It resets host rules and added back the rule for GHE server but the rule to handle the Artifact Registry repository was not added back and this is the point where NPM lookups started to fail with HTTP 403.

```
"Reinitializing hostRules for repo","time":"2023-05-29T16:22:14.276Z"
"Clearing hostRules","time":"2023-05-29T16:22:14.276Z"
"Adding token authentication for my-ghe-server.com to hostRules"
```

[Full renovate.log](https://github.com/renovatebot/renovate/files/11593149/renovate.log)

[config.js](https://github.com/renovatebot/renovate/files/11593248/config-obfuscated.js.txt)

**Comment by rvillane** [comment]
Seems like if this comparison is eliminated form the `setNpmrc()` function, the npmrc auth settings are restored for each repo

```
    if (input === npmrcRaw) {
         return;
    }
```
https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/npm/npmrc.ts#L127-L129

Feel like this comparison was an optimization added some time ago, is not a recent change, and it tries to save the cost of setting the npmrc configuration again when it supposedly has not changed but in my case is what prevents npmrc settings from being restored when each repo hostRules are cleared out and initialized.

thoughts ?

**Reply by rarkins** [reply]
Thoughts:
1. `npmrcRaw` perhaps needs resetting per-repo, and that's the true root cause
2. If that's to hard to do, maybe we can remove this optimization if it's not saving much

**Comment by rvillane** [comment]
resetting `npmrcRaw` and removing the optimization will achieve the same result, setting the npmrc settings at the beginning of every repo scanning, I think.  What I mean is that resetting `npmrcRaw` on every repo just invalidates the purpose of the original optimization.

**Reply by rarkins** [reply]
I think the optimization also exists for the purpose of same npmrc used multiple times in the same repo? You could add a log statement to see how many times it's called in yours, for example

**Reply by rvillane** [reply]
yup, I did and this setNpmrc function is called only once at the beginning of every repo scanning.

**Reply by rarkins** [reply]
<img width="457" alt="image" src="https://github.com/renovatebot/renovate/assets/6311784/e5d73f84-63ac-4436-b14f-a3b7ba7423fa">

That first one definitely might get called often. Maybe a safer way would be to call it twice at the start of repo - empty and then the existing one. That way the second call always has an effect, and we don't need to worry about the other cases for now

**Reply by rvillane** [reply]
You definitely know the tool better than I do and is probably better if you make this code change since I got a bit lost in your explanation above but I can definitely test it once the fix is available

**Reply by rarkins** [reply, marked as answer]
https://github.com/renovatebot/renovate/pull/22515

You can maybe make this simple edit locally and test already

**Reply by rvillane** [reply]
@rarkins, I just tried that change locally and can confirm it works as expected

**Comment by rakins** [comment]
Please remove me from this email distribution!!!!!
    On Tuesday, May 30, 2023 at 11:49:48 PM PDT, Rhys Arkins ***@***.***> wrote:  
 
 


#22515

You can maybe make this simple edit locally and test already
